subprojects/docs/src/docs/userguide/javaPlugin.adoc

```diff
@@ -520,11 +525,18 @@ You can then register it for incremental compilation using the `getSupportedOpti
 </sample>
 ++++
 
-Processors that don't fit these categories will result in full recompilation.
-This includes processors that use `java.io` instead of the `Filer` API and processors that need to read or write resource files.
+Both categories have the following limitations:
+
+* Must not have any side effects except for using the `Filer` and `Messager` APIs.
+For instance, writing files using `java.io` will result in silent failures later on, as these files won't be cleaned up correctly.
+If your processor does this, it cannot be incremental.
+* Must not depend on compiler-specific APIs like `com.sun.source.util.Trees`. Gradle wraps the processing APIs, so attempts to cast to compiler-specific types will fail.
```

Comment:
I would prefix each line with "They (must not ...)", otherwise the subject is a little ambiguous. It would also be nice if you could clarify what you mean by "side effects" in this instance. I'm guessing you mean changes outside of the JVM's memory, but I'm not sure.

Reply:
Similar to above I'm not sure how to phrase this better without being too long-winded. An annotation processor runs as part of compilation and could contain arbitrary code. As such it could have arbitrary side effects such as writing anywhere in the file system, making network calls or formatting your hard drive. Processors shouldn't do this of course, they should only use the ProcessingEnvironment API.

Reply:
Perhaps: They must not affect external systems, such as the network or file system, except through the \`Filer\` and \`Messager\` APIs.

Alternatively, but a more wordy: They must not affect the environment outside of the JVM, such as the network or file system, except through the \`Filer\` and \`Messager\` APIs.

Also, would it be possible to link to those APIs?